### PR TITLE
Implement the BOM removal and rails rake task hook

### PR DESCRIPTION
**Pull request by h-lame**
For: https://trello.com/c/UcJJBH0k/183-investigate-bom-issue-with-asset-compilation

We look at each css file in the rails asset location and if we detect a
BOM we remove it from the file and also regenerate the gzipped version
(if it exists) using the new BOM-less version of the CSS file (using the
same bit of sprockets that rails uses to do so).

It leaves files without a BOM file alone, and leaves files that are not
CSS files alone too.

We hook into rails by enhancing the `assets:precompile` task to invoke
ours once it has completed.

### CODE REVIEW

**Comment by NickColley** [comment]
This is so good 😍

File: spec/fixtures/no-bom-in-here.css [hunk 1]
```diff
@@ -0,0 +1 @@
+h1 { color: purple; font-weight: bolder; font-family: "Comic Sans", fantasy; }
```
Comment (fofr) [inline spec/fixtures/no-bom-in-here.css:1]: Best fixtures.

**Comment by fofr** [comment]
Really pleased to see this. Leaving for someone with a bit more indepth Rails knowledge to review.

Probably worth putting prominently in the Readme what a BOM is.

**Comment by h-lame** [comment]
One downside to this is that it processes all the css files in the public folder.  We seem to leave old assets lying around which means we process them too and remove the BOM.  On integration I deployed with SRI first to check it was broken in FF51 and then again with this tool and it broke the SRI sha of the cached page in chrome.  I think because the HTML page had a sha calculated based on the BOM version of the CSS - when we removed the BOM the sha changes, but the HTML doesn't take that into account as it's cached.

This probably won't be a problem for us as we'll deploy with SRI + BOM removal in one go, but we might need to work out which assets to deal with in some other way as not everyone will have the same deploy stack.  It generally feels wrong to mutate assets we didn't just generate too.

**Comment by h-lame** [comment]
Ok.  Closing this for now because stripping the BOM after the fact doesn't work.  When compiling the assets rails pre-calculates a bunch of stuff and writes that into a manifest file.  In production mode rails will read the manifest file instead of loading the asset and calculating things on the fly.  This is obviously for performance reasons and very sensible.

One of the things in the manifest file is the digest of the file contents, and it is the digest of the file contents that is used to calculate the SRI checksum that is added to the script and link tags.  See [`asset_integrity`](https://github.com/rails/sprockets-rails/blob/v3.1.0/lib/sprockets/rails/helper.rb#L107-L113) in sprockets-rails which ultimately calls [`Manifest#integrity`](https://github.com/rails/sprockets-rails/blob/v3.1.0/lib/sprockets/rails/helper.rb#L279-L283) which reads the intergrity value out of the manifest metadata (see: [`Sprockets::Manifest`](https://github.com/rails/sprockets/blob/v3.7.0/lib/sprockets/manifest.rb)).

If we strip the BOM from the CSS files after the manifest file has been written then we've changed the file contents which means the digest would change which means the SRI checksum should change, but as we've already written the manifest file at this point it won't.

I think we're going to need an alternative approach where we try to hook directly into the sass processor and strip the BOM from the file after Sass has done its thing, but before it is passed fully back to sprockets and rails for manifest generation.

**Pull request closed without merging**